pageserver/src/layered_repository/layer_map.rs

```diff
@@ -53,14 +57,138 @@ pub struct LayerMap {
     pub frozen_layers: VecDeque<Arc<InMemoryLayer>>,
 
     /// All the historic layers are kept here
+    historic_layers: RTree<LayerEnvelope>,
 
-    /// TODO: This is a placeholder implementation of a data structure
-    /// to hold information about all the layer files on disk and in
-    /// S3. Currently, it's just a vector and all operations perform a
-    /// linear scan over it.  That obviously becomes slow as the
-    /// number of layers grows. I'm imagining that an R-tree or some
-    /// other 2D data structure would be the long-term solution here.
-    historic_layers: Vec<Arc<dyn Layer>>,
+    /// L0 layers has key range: (Key::MIN..Key::MAX) and locating them using R-Tree search is very inefficient.
+    /// So l) layers are also pushed in l0_layers vector.
+    l0_layers: Vec<Arc<dyn Layer>>,
+}
+
+struct LayerEnvelope {
+    layer: Arc<dyn Layer>,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Debug)]
+struct IntKey(i128);
```

Comment:
How about using i128 directly? All the traits,`Bounded`, `Signed`,  etc. are already implemented for i128.

Reply:
Sorry, the problem is with area calculation. `rstar` doesn't try to perform "wide"  multiplication and so result of calculation may be incorrect because of overflow. This is why we can not use i128 directly.

I have made many different experiments with solving overflow problem.
The straightforward attempt to use BitInt failed because it is not implementing `Copy`.
Hacker's approach is to calculate sqrt(a*b) instead of a*b. It ca work because multiplication is used by `rstar` only for area calculation. We can covert i128 operands to BigInt, multiply two BigInts, take square and store it back as i128.
But tests doesn't show some performance improvement of such approach comparing with just wrapper multiplication.

Reply:
> Sorry, the problem is with area calculation. `rstar` doesn't try to perform "wide" multiplication and so result of calculation may be incorrect because of overflow. This is why we can not use i128 directly.

I don't understand. Is the problem that `int128 * int128` might overflow? But you're implementing `Mul` by calling `wrapping_mul`, so that will overflow too. Or is the problem that if `int128 * int128` might overflow *and panic*, and you want to avoid the panic? But wrapping around and returning incorrect result doesn't seem nice either.

Is this a problem just for `i128`, or would `i64` or `i32` have the same overflow problem?

> I have made many different experiments with solving overflow problem. The straightforward attempt to use BitInt failed because it is not implementing `Copy`. Hacker's approach is to calculate sqrt(a_b) instead of a_b. It ca work because multiplication is used by `rstar` only for area calculation. We can covert i128 operands to BigInt, multiply two BigInts, take square and store it back as i128. But tests doesn't show some performance improvement of such approach comparing with just wrapper multiplication.

Reply:
> Is this a problem just for `i128`, or would `i64` or `i32` have the same overflow problem?

The main problem is certainly panic caused by overflow.
It takes place for all integer types, because `rstar` is not supporting widening multiplication (i.e. i64*i64->i128).
 
Yes, in case of wrapping multiplication calculated area may be not correct.
But it just cause non-optimal R-Tree. It should not cause lookup failures.

More precise calculation of area is the following:
```
impl Mul for IntKey {
    type Output = Self;
    fn mul(self, rhs: Self) -> Self::Output {
	let a = BigUint::from(self.0 as u128);
	let b = BigUint::from(rhs.0 as u128);
        IntKey((a * b).sqrt().to_u128().unwrap() as i128)
    }
}
```
Certainly it also may be not absolutely correct because of rounding.
I did some tests but can not find meaningful difference (in time) between iplementation with wrapping multplication and using BigInt. In means two different things:
1. BigInt arithmetic is not so expensive (comparing with other operations).
2. Non-optimal tree built because of incorrect area calculation is not so different from optimal (at least at the example I have used)

Reply:
Ok. `saturating_mul` would probably be less wrong than `wrapping_mul`. 

Please explain all of this in the comments.

Reply:
> Ok. `saturating_mul` would probably be less wrong than `wrapping_mul`.
> 
> Please explain all of this in the comments.

I decided that approach with using BitInt for multiplication is more correct, so I have implemented it and add comments.